src/store/concepts/session/operations/__tests__/login.spec.js

```diff
@@ -0,0 +1,129 @@
+import {
+  USER_LOGIN_SUBMIT,
+  USER_LOGIN_SUCCESS
+} from 'Store/concepts/session/types'
+
+import loginUserOperation from 'Store/concepts/session/operations/login'
+import storeWithMiddlewareMock from 'Store/__mocks__/storeWithMiddlewareMock'
+import mockHttpClient from 'Api/__mocks__/mockHttpClient'
+import storage from 'Utils/storage'
+
+jest.mock('Utils/storage')
+
+describe('loginUserOperation()', () => {
+  storage.session.set.mockImplementation(jest.fn())
+
+  describe('with success response', () => {
+    const formMocks = {
+      form: { setStatus: jest.fn(), setSubmitting: jest.fn() }
+    }
+    const httpClient = mockHttpClient([
+      {
+        method: 'get',
+        resolve: { data: { request_token: 'test' } }
+      },
+      {
+        method: 'post',
+        resolve: { data: { session_id: 'testId' } }
+      }
+    ])
+    const { store, logicMiddleware } = storeWithMiddlewareMock(httpClient, [
+      loginUserOperation
+    ])
+
+    store.dispatch({
+      type: USER_LOGIN_SUBMIT,
+      values: { username: 'test', password: 'test' },
+      ...formMocks
+    })
+
+    it('handles USER_LOGIN_SUCCESS', async () => {
+      await logicMiddleware.whenComplete()
+
+      expect(store.getActions()).toStrictEqual([
+        {
+          type: USER_LOGIN_SUBMIT,
+          values: {
+            username: 'test',
+            password: 'test'
+          },
+          ...formMocks
+        },
+        {
+          type: USER_LOGIN_SUCCESS
+        }
+      ])
+      expect(storage.session.set).toHaveBeenCalledTimes(1)
+    })
+  })
+
+  describe('with errors', () => {
+    it('in response', async () => {
+      const formMocks = {
+        form: { setStatus: jest.fn(), setSubmitting: jest.fn() }
+      }
+      const error = new Error()
+      error.status = 401
+      const httpClient = mockHttpClient([
+        {
+          method: 'get',
+          reject: error
+        }
+      ])
+      const { store, logicMiddleware } = storeWithMiddlewareMock(httpClient, [
+        loginUserOperation
+      ])
+
+      store.dispatch({
+        type: USER_LOGIN_SUBMIT,
+        values: { username: 'test', password: 'test' },
+        ...formMocks
+      })
+      await logicMiddleware.whenComplete()
+
+      expect(store.getActions()).toStrictEqual([
+        {
+          type: USER_LOGIN_SUBMIT,
+          values: {
+            username: 'test',
+            password: 'test'
+          },
+          ...formMocks
+        }
+      ])
+    })
+
+    it('login action with incorrect error', async () => {
+      const formMocks = {
```

Comment:
fromMock

Reply:
can we place it on top and reuse in each test case?

Reply:
No, we have different mocks for different cases.

Reply:
you can use something like this to reset mocks between tests:
```
beforeEach(() => {
  jest.clearAllMocks();
});
```